### PR TITLE
fix #19684 barline mmrest crash

### LINKED ISSUE #19684: Crash after adding some measures and changing repeat barline (with enabled multimeasure rest mode)
### Issue type

Crash or freeze

### Bug description

_No response_

### Steps to reproduce

1. Create template
2. Add repeat barline to the beginning of 2nd measure
3. Press M (Toggle multimeasure rest)
4. Select 1st measure 
5. Insert some measures **After selection** 
6. Select repeat barline
7. Change it to normal barline 

### Screenshots/Screen recordings

https://github.com/musescore/MuseScore/assets/90187801/fcd811a1-ee13-480d-b722-5a4dc65298ed

### MuseScore Version

4.2.0 master

### Regression

Yes, this used to work in a previous version of MuseScore 4.x

### Operating system

Windows10

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sammik**
Resolves: #19684 

End barline is added to measure during layout. It means, in some situations, if we are in mmrest mode and measure was not in score before in standard mode, it has no end barline.

Other situation, with exactly same problem is:
1. create empty score
2. create part
3. add start repeat somewhere to score
4. switch to part
5. select start repeat barline and change it to some other one

So if we are in mmrest mode, we need use mmrest measures and their barlines (as standard barlines may be undefined).

In first commits, there are just nullchecks.
Second commit handles barlines and measures in mmrest mode correctly.

TODO: there are still weired places in code (and coresponding issues), so it needs some refactor, but may be, I can post them as separate issue.

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [x] I created a unit test or vtest to verify the changes I made (if applicable)

### CODE REVIEW

**Comment by sammik** [comment]
@Eism

**Comment by sammik** [comment]
Now I realized, that I forgot to correct lines 150 - 152

I will correct it may be tonight.

**Comment by MarcSabatella** [comment]
FWIW, I worked on this code a bit years ago and do remember some things.  I definitely recall bugs over the years that resulted from the fact that either the underlying measures or the mmrests are laid out on score load,  the but not both.  So yes, the parent (system) can be null.  I always figured it would possible to assign *something* on load to avoid the worst of the bugs (null deference crashes), but there would still be plenty of other places for things to go awry.

File: src/engraving/dom/barline.cpp [hunk 1]
```diff
@@ -130,7 +147,7 @@ static void undoChangeBarLineType(BarLine* bl, BarLineType barType, bool allStav
                     }
 
                     lmeasure->undoChangeProperty(Pid::REPEAT_END, false);
-                    if (lmeasure->nextMeasure() && !lmeasure->nextMeasure()->isFirstInSystem()) {
+                    if (lmeasure->nextMeasure() && (!lmeasure->nextMeasure()->system() || !lmeasure->nextMeasure()->isFirstInSystem())) {
```
Comment (Eism) [inline src/engraving/dom/barline.cpp:150]: Please move lmeasure->nextMeasure() to a variable and reuse it here and below

**Comment by sammik** [comment]
@Eism One question.

I have almost done fixes for few other issues regarding barline changes.

Should I push it here, or as separate PR?

Thank You

**Comment by Eism** [comment]
@sammik 
As separate PR please :)

**Comment by sammik** [comment]
@Eism done

**Review by Eism — APPROVED** [review]

**Comment by DmitryArefiev** [comment]
Tested #19684 on Win10, Mac14, LinuxUbuntu 22.04 - **FIXED**

Also, found some bugs related to repeat barline change (not a recent regressions) #19776 #19778

**Pull request merged**